### PR TITLE
Implement writing pandas metadata and auto-setting cats/index

### PR DESCRIPTION
**Pull request by martindurant**
Only copes with single index, not compound (as was the case in fastparquet
so far).

See the [spec](https://github.com/pandas-dev/pandas/blob/master/doc/source/developer.rst)

### CODE REVIEW

File: fastparquet/test/test_output.py [hunk 1]
```diff
@@ -381,15 +381,25 @@ def test_write_compression_schema(tempdir):
...
                        index=pd.Index([10, 20, 30], name='z'))
 
     writer.write(fn, df)
 
-    r = ParquetFile(fn)
-    assert set(r.columns) == {'x', 'y', 'z'}
+    pf = ParquetFile(fn)
+    assert set(pf.columns) == {'x', 'y', 'z'}
+    meta = json.loads(pf.key_value_metadata['pandas'])
+    assert meta['index_columns'] == ['z']
```
Comment (wesm) [inline fastparquet/test/test_output.py]: So one deficiency for named indexes is that there may be a conflict between the index name and a column name in the DataFrame. I opened https://github.com/pandas-dev/pandas/issues/16391, which is something we can add in a forward-compatible way

File: fastparquet/writer.py [hunk 2]
```diff
@@ -671,6 +678,7 @@ def make_metadata(data, has_nulls=True, ignore_columns=[], fixed_text=None,
             se.repetition_type = parquet_thrift.FieldRepetitionType.OPTIONAL
         fmd.schema.append(se)
         root.num_children += 1
+    meta.value = json.dumps(pandas_metadata)
     cats.value = json.dumps(catstruct)
```
Comment (wesm) [inline fastparquet/writer.py]: Do you want to drop this one in favor of the pandas metadata?

File: fastparquet/test/test_output.py (hunk 1 above)
Comment (martindurant) [inline fastparquet/test/test_output.py]: We currently check against duplicate column names after reset_index, so you would get an error for this case

File: fastparquet/writer.py (hunk 2 above)
Comment (martindurant) [inline fastparquet/writer.py]: You are right, this should be removed. On reading I give a warning when this old version is present.

**Pull request merged**